File: api/middleware/middleware.go
package middleware

import (
	"bytes"
	"io"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/solocoder/session147/internal/common/logger"
	"github.com/solocoder/session147/internal/common/model"
	"go.uber.org/zap"
)

func CORSMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")
		c.Writer.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
		c.Writer.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

		if c.Request.Method == "OPTIONS" {
			c.AbortWithStatus(204)
			return
		}

		c.Next()
	}
}

func RequestLogger() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()
		path := c.Request.URL.Path
		query := c.Request.URL.RawQuery

		var body []byte
		if c.Request.Body != nil {
			body, _ = io.ReadAll(c.Request.Body)
			c.Request.Body = io.NopCloser(bytes.NewBuffer(body))
		}

		c.Next()

		latency := time.Since(start)
		status := c.Writer.Status()
		clientIP := c.ClientIP()
		method := c.Request.Method

		logger.Info("HTTP Request",
			zap.Int("status", status),
			zap.String("method", method),
			zap.String("path", path),
			zap.String("query", query),
			zap.String("ip", clientIP),
			zap.Duration("latency", latency),
		)
	}
}

func AuthMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			c.JSON(401, model.ApiResponse{
				Code:    401,
				Message: "authorization header is required",
			})
			c.Abort()
			return
		}

		c.Set("user_id", "anonymous")
		c.Next()
	}
}

func ErrorHandler() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Next()

		if len(c.Errors) > 0 {
			err := c.Errors.Last()
			logger.Error("Request error", zap.Error(err.Err))

			code := 500
			message := "Internal Server Error"

			c.JSON(code, model.ApiResponse{
				Code:    code,
				Message: message,
			})
		}
	}
}

func RateLimiter() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Next()
	}
}

File: src/modules/gateway/middleware.ts
import { Request, Response, NextFunction } from 'express';
import type {
  IAuthService,
  IRateLimiter,
  IRequestContext,
} from '@ports/index';
import { generateTraceId } from '@utils/index';
import { rootLogger } from '@modules/logging';
import type { UserPrincipal } from '@apptypes/index';

declare global {
  namespace Express {
    interface Request {
      context: IRequestContext;
      principal?: UserPrincipal;
    }
  }
}

const AUTH_HEADER_PREFIX = 'Bearer ';
const RATE_LIMIT_HEADER_MAX = 100;

interface ErrorResponse {
  code: number;
  message: string;
}

export const createRequestContext = (
  req: Request,
  _res: Response,
  next: NextFunction,
): void => {
  req.context = {
    traceId: extractTraceId(req),
    spanId: generateTraceId().slice(0, 16),
    principal: null,
    startTime: Date.now(),
    attributes: {},
  };
  next();
};

export const createAuthMiddleware = (authService: IAuthService) => {
  return async (
    req: Request,
    res: Response,
    next: NextFunction,
  ): Promise<void> => {
    const token = extractBearerToken(req);
    if (!token) {
      sendErrorResponse(res, 401, 'Missing or invalid authorization header');
      return;
    }

    const principal = await authService.authenticate(token);
    if (!principal) {
      sendErrorResponse(res, 401, 'Invalid or expired token');
      return;
    }

    req.principal = principal;
    req.context.principal = principal;
    next();
  };
};

export const createPermissionMiddleware = (
  authService: IAuthService,
  permission: string,
) => {
  return (req: Request, res: Response, next: NextFunction): void => {
    if (!req.principal) {
      sendErrorResponse(res, 401, 'Authentication required');
      return;
    }

    if (!authService.authorize(req.principal, permission)) {
      sendErrorResponse(res, 403, 'Insufficient permissions');
      return;
    }

    next();
  };
};

export const createRateLimitMiddleware = (rateLimiter: IRateLimiter) => {
  return async (
    req: Request,
    res: Response,
    next: NextFunction,
  ): Promise<void> => {
    const key = buildRateLimitKey(req);
    const result = await rateLimiter.checkLimit(key);

    setRateLimitHeaders(res, result);

    if (!result.allowed) {
      setRetryAfterHeader(res, result.resetTime);
      sendErrorResponse(
        res,
        429,
        'Rate limit exceeded. Please try again later.',
      );
      return;
    }

    next();
  };
};

export const requestLogger = (
  req: Request,
  res: Response,
  next: NextFunction,
): void => {
  const start = Date.now();

  res.on('finish', () => {
    logRequest(req, res, start);
  });

  next();
};

export const errorHandler = (
  err: Error,
  req: Request,
  res: Response,
  _next: NextFunction,
): void => {
  rootLogger.error('Unhandled error', {
    trace_id: req.context?.traceId,
    error: err.message,
    stack: err.stack,
    path: req.path,
    method: req.method,
  });

  sendErrorResponse(res, 500, 'Internal server error');
};

function extractTraceId(req: Request): string {
  return (req.headers['x-trace-id'] as string) || generateTraceId();
}

function extractBearerToken(req: Request): string | null {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith(AUTH_HEADER_PREFIX)) {
    return null;
  }
  return authHeader.slice(AUTH_HEADER_PREFIX.length);
}

function buildRateLimitKey(req: Request): string {
  const clientId =
    (req.headers['x-client-id'] as string) ||
    req.principal?.user_id ||
    req.ip ||
    'unknown';
  return `${clientId}:${req.method}:${req.path}`;
}

function setRateLimitHeaders(
  res: Response,
  result: { remaining: number; resetTime: number },
): void {
  res.setHeader('X-RateLimit-Limit', String(RATE_LIMIT_HEADER_MAX));
  res.setHeader('X-RateLimit-Remaining', String(result.remaining));
  res.setHeader('X-RateLimit-Reset', String(Math.ceil(result.resetTime / 1000)));
}

function setRetryAfterHeader(res: Response, resetTime: number): void {
  const retryAfter = Math.ceil((resetTime - Date.now()) / 1000);
  res.setHeader('Retry-After', String(retryAfter));
}

function sendErrorResponse(
  res: Response,
  code: number,
  message: string,
): void {
  const response: ErrorResponse = { code, message };
  res.status(code).json(response);
}

function logRequest(req: Request, res: Response, start: number): void {
  const duration = Date.now() - start;
  const logger = rootLogger.child({
    trace_id: req.context?.traceId,
    span_id: req.context?.spanId,
  });

  const logData = {
    method: req.method,
    path: req.path,
    status_code: res.statusCode,
    duration_ms: duration,
    user_id: req.principal?.user_id,
    client_ip: req.ip,
    user_agent: req.headers['user-agent'],
  };

  const level = getLogLevel(res.statusCode);
  logger[level]('Request completed', logData);
}

function getLogLevel(statusCode: number): 'error' | 'warn' | 'info' {
  if (statusCode >= 500) return 'error';
  if (statusCode >= 400) return 'warn';
  return 'info';
}
